From david-b@pacbell.net Fri Apr 18 18:59:16 2008
From: David Lopo <lopo.david@gmail.com>
Date: Fri, 18 Apr 2008 18:49:01 -0700
Subject: usb: gadget zero timer init fix
To: Greg KH <greg@kroah.com>
Cc: linux-usb@vger.kernel.org, David Lopo <lopo.david@gmail.com>
Message-ID: <200804181849.02001.david-b@pacbell.net>
Content-Disposition: inline


From: David Lopo <lopo.david@gmail.com>

Initialize timer earlier so if an error occurs allocating USB request
or buffer request (zero_bind) Gadget Zero will not hang trying to
delete an uninitialized timer (zero_unbind).

Signed-off-by: David Lopo <lopo.david@gmail.com>
Signed-off-by: David Brownell <dbrownell@users.sourceforge.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/gadget/zero.c |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

--- a/drivers/usb/gadget/zero.c
+++ b/drivers/usb/gadget/zero.c
@@ -1134,6 +1134,10 @@ autoconf_fail:
 	dev->gadget = gadget;
 	set_gadget_data(gadget, dev);
 
+	init_timer(&dev->resume);
+	dev->resume.function = zero_autoresume;
+	dev->resume.data = (unsigned long) dev;
+
 	/* preallocate control response and buffer */
 	dev->req = usb_ep_alloc_request(gadget->ep0, GFP_KERNEL);
 	if (!dev->req)
@@ -1165,9 +1169,6 @@ autoconf_fail:
 
 	usb_gadget_set_selfpowered(gadget);
 
-	init_timer(&dev->resume);
-	dev->resume.function = zero_autoresume;
-	dev->resume.data = (unsigned long) dev;
 	if (autoresume) {
 		source_sink_config.bmAttributes |= USB_CONFIG_ATT_WAKEUP;
 		loopback_config.bmAttributes |= USB_CONFIG_ATT_WAKEUP;
